lucene/highlighter/src/java/org/apache/lucene/search/uhighlight/LengthGoalBreakIterator.java

```diff
@@ -174,7 +175,48 @@ private int moveToBreak(int idx) { // precondition: idx is a known break
   // called at start of new Passage given first word start offset
   @Override
   public int preceding(int offset) {
-    return baseIter.preceding(offset); // no change needed
+    final int fragmentStart = Math.max(baseIter.preceding(offset), 0); // convert DONE to 0
+    fragmentEndFromPreceding = baseIter.following(fragmentStart);
+    if (fragmentEndFromPreceding == DONE) {
+      fragmentEndFromPreceding = baseIter.last();
+    }
+    final int centerLength = fragmentEndFromPreceding - fragmentStart;
+    final int extraPrecedingLengthGoal = (int)((lengthGoal - centerLength) * fragmentAlignment);
```

Comment:
I'm noticing that the logic here for fragment alignment doesn't seem to care whatsoever about where "offset" is within it's segment.  But isn't that super relevant?  As I consider a BI on SENTENCE based segments, I think it is... but I can see how you overlooked this when focusing on WORD scenarios.  For example assuming 0.5 fragmentAlignment, and if the "offset" (the match) happens to occur at the right end of the segment, and lets say the length goal is only 10 chars larger than centerLength, then shouldn't we expand to the right and not the left?

You're going to hate me for this but let me make a proposal :-)  What if we multiply fragmentAlignment by lengthGoal and interpret this as a minimum number of characters wanted to the left of the start of the match (`offset`).  The difference of that with lengthGoal indicates minimum chars wanted to the right.  We use the delegate BI to find the passage start, and then we can consult fragmentAlignment with where `offset` is relative to the start to decide how much text to the right of `offset` we want.

Reply:
On the contrary!

This suggestion is a little different than what I've thought of, so I'll have to rework the whole thing, but it has a great quality: This would make the length of the match irrelevant for the GLBI logic. I'm not going to go through all the details, but the code can be a little more optimal and less... hacky.

I should have read this before writing [my previous comment](https://github.com/apache/lucene-solr/pull/1123#discussion_r361529545), now it is deprecated.

Reply:
I'm glad you're cool with entertaining reworking the whole thing :-)  BTW my proposal is rather incomplete or imperfect or both.  We probably won't get a perfect lengthGoal hit (hence "goal" in the name).  And we almost certainly won't get a perfect fragmentAlignment ratio with before/after -- it should perhaps also have "goal" in the name.  So whatever the solution is will have some inherent fuzziness to it.  This is probably all obvious to you but I just want to state I'm aware you can poke lots of holes in my proposal.  I just think we can do a bit better than your original proposal by using some more information.

Reply:
BTW see Solr's "LuceneRegexFragmenter" which has a "slop" notion and may have ideas worth looking at.

Reply:
I have a working version with this new behavior. There is just a single assertion that fails [here](https://github.com/apache/lucene-solr/blob/e06ad4cfb587e1bb69a80c0b531df0e52c2da89b/solr/core/src/test/org/apache/solr/highlight/TestUnifiedSolrHighlighter.java#L267). It's because the `fragsize=60` there means to try and get 60 characters on the right side of the match as the default `fragalign` is zero. To pass the test the query parameters need to be changed to have `hl.fragsize=50&hl.fragalign=0.5`.
This is not a problem by itself, but there is no way to make the new behavior completely backwards-compatible like the previous one.

It could also be a bit jarring to have `fragsize` with a distinctively different meaning compared to the other highlighters:
- For original&fastVector the `fragsize` is the target size of the snippet.
- For unified the `fragsize` is the size of the contextual text around the match. Ie.: the snippet's target size is `fragsize+matchsize`.

I'm still in favor of this change, there's literally half as many calls to `preceding()` and `following()` of the underlying BI and it's much clearer as well.

Should I change the failing test and push this for review? What about the backward-compatibility? I'm not sure how to change the docs or the default parameters.

Reply:
Correction: the snippet size is not `fragsize+matchsize`, but `min(sentence_left_of_match, fragsize*fragalign) + matchsize + min(sentence_right_of_match, fragsize*(1-fragalign))`.
So if I want a left aligned highlight in SENTENCE mode, but the match is on the right end of the sentence, the whole `fragsize` would be added on the right.
This is less accurate if you're concerned about the result snippet's size, but more accurate if you're concerned about the alignment of the match.

Reply:
When WORD mode is used, is there no change RE fragsize interpretation?  By "matchsize" do you mean the length of the matched words themselves?

Reply:
The change is the same in WORD mode too, but the match and the fragment of the match is mostly the same. It's just not that apparent compared to the SENTENCE mode.

I'll try and clarify all the expressions:
- Source text: `I'm looking for a beach ball for my vacation. I've lost the old one. I liked that.`
- Query text: `vacation`
- Match term is `Match{start=36, end=44, text=vacation}` thus `matchsize` is `8`
- Match-fragment is the fragment that contains the match according to the underlying BI:
  - WORD: `vacation` because it breaks on space and period.
  - SENTENCE: `I'm looking for a beach ball for my vacation.` because it breaks on period.
- `hl.bs.type=SENTENCE&hl.fragsize=10&hl.fragalign=0.0`:
  - Match fragment `I'm looking for a beach ball for my vacation.` has 1 extra char on the right, but 10 is requested. Will try to append next fragments to reach `match.end+((1-fragalign)-fragsize)` (which is `54`).
  - Next break after `54` is the end of the second sentence so the result snippet will be: `I'm looking for a beach ball for my vacation. I've lost the old one.`
- `hl.bs.type=SENTENCE&hl.fragsize=10&hl.fragalign=0.5`:
  - Match fragment `I'm looking for a beach ball for my vacation.` has 36 extra chars on the left, while 5 is requested. Also 5 are requested on the right so the extension will happen just like in the previous example. The result will be `I'm looking for a beach ball for my vacation. I've lost the old one.`.
- `hl.bs.type=SENTENCE&hl.fragsize=10&hl.fragalign=1.0`:
  - Match fragment `I'm looking for a beach ball for my vacation.` has 36 extra chars on the left, while 10 is requested. No extra chars are requested on the right so the match fragment is the result.

Reply:
Thanks for the detailed example.

> As you can see the fragsize is not about the final size of the snippet, but about the text around the match.

I think it's a bit strong to claim fragsize is _not_ about the final size of the snippet.  There is a linear relationship.  Decreasing this reduces snippet sizes and increasing it will increase snippet sizes.  It's just rather indirect because we are prioritizing the alignment.  And perhaps we overshoot too much.

Also; couldn't we subtract out the match length itself if it's unduly influencing the final snippet size?

Lets pretend we're at a more perfect future and not concerning ourselves with backwards compatibility.  Realistically, do you imagine a user would want to use fragalign of anything other than 0.5?  I doubt it; I think we should assume users want this.  Advanced users can choose _not_ to use this BreakIterator -- there are others plus they can write their own (which I've done).  Also, rather than assuming users need 50% of fragsize (lengthGoal) on both sides, maybe assume 60% _of that_ (I'm looking at default slop in LuceneRegexFragmenter which is similar), so thus ~0.3 (or perhaps 1/3rd) fragsize minimum left & right.  I think this will lead to snippets that are both closer to the desired fragsize on average and lead to more balanced passages than we have today.  LMK what you think.

Reply:
Maybe one problem with this is that it wouldn't be quite as ideal for WORD, which would in practice always put the word at the 1/3rd and not half way.  This generic wrapping BI doesn't know on average how big the underlying BI will tend to be (SENTENCE vs WORD).  I suppose that could be addressed by making the "slop" configurable.

Reply:
> Also; couldn't we subtract out the match length itself if it's unduly influencing the final snippet size?

My original patch was so convoluted because that cannot be done easily in the current implementation with the BreakIterator base class. The call to `preceding()` on the LGBI receives the `matchStart` and returns the final start of the snippet. Since this method has no way to know the other end of the match it cannot adjust to the length of the match. Using private state the `following()` could know it, but that  would only be able to adjust on the right hand side and not the left.
Something like `public int[] getPassageRange(int matchStart, int matchEnd) { ... }` would allow a better implementation to be written. But that requires changing `protected final BreakIterator breakIterator;` inside `FieldHighlighter` to a custom BI base that has such method.

> Realistically, do you imagine a user would want to use fragalign of anything other than 0.5?

I'd say the margin values of `fragalign` are of little use, but having the ability to set it in the range of`0.35-0.65` could be worthwhile.

I'm not sure if a `slop` parameter similar to RegexFragmenter's would be too beneficial.
- The `RegexFragmenter` processes the whole stored text (until `hl.regex.maxAnalyzedChars`) and creates a list of breaks called `hotSpots`. Interestingly it also uses the tokenizer's offsets from the index (I think), because even if a break would be in the middle of a word it will not break it in half. Unless I'm completely missing something the regex seems to only give a strong recommendation to where a break should go.
- Implementing an acceptable range (`slop`) would require to build some list of `hotSpots` or at least to look for them when extracting the fragment. For this we could only use `preceding()` and `following()` on the wrapped BI, which we wanted to avoid if possible for performance reasons.
- For WORD BI the current solution is pretty accurate and is as optimal as it gets. (single call of `preceding()` and `following()` on the underlying BI)
- For SENTENCE BI the slop would not help enough to be worth the trouble I think. Let's say I have to extend the snippet on the right side with 100 chars of a slop 0.4 (valid range: 60-140):
  - If the next sentence is 300 chars long there's nothing better to do what we already do.
  - If the next sentences have the lengths of [30, 50, 200, 40] the better choice would be to have the first two sentences appended with a sum length of 80. (that's only 20 off from the goal) For this the LGBI can be created with `closestTo` mode instead of  `minLength` mode. It would cost an extra call of `preceding()` and `following()` each but would produce the better result instead of appending 3 of the following sentences with 280 total length.

So we already have a `slop` of sorts... it's just a boolean that cannot be turned on right now as it is missing a parameter. I've mentioned [this todo](https://github.com/apache/lucene-solr/blob/e06ad4cfb587e1bb69a80c0b531df0e52c2da89b/solr/core/src/java/org/apache/solr/highlight/UnifiedSolrHighlighter.java#L330) about it before.

After looking though all of this, my suggestion is to:
- keep `hl.fragalign` and have it be `0.5` by default
- add `hl.closestTo` with default `false`
  - a better parameter name would be welcome
  - we could consider making it automatically turned on if the BI is SENTENCE
- do not worry about keeping a backward-compatible behavior with this change

Reply:
I just want to confirm: your suggestion will be the code you already wrote above but have not yet pushed and that breaks that one test.  Yes. we can just change that test slightly.

One thing that vexes me a bit about highlighting is that there are so many choices and options to name -- configuration-itis.  Good defaults help.  Also simply making a reasonable choice for 98% of users and letting the 2% do something themselves is also better than adding yet another tuning nob.  I'm not sure where the things we are discussing lie.  (Oh and deprecate/removing the other highlighters would help too ;-)  Any way...

> hl.fragalign

Since you propose the default be 0.5 then in what scenario can you imagine someone would specify it?  I suggest `hl.fragAlignRatio` as a better name if we keep this.

> hl.closestTo

I wish it simply always worked this way (`true`), to be honest.  If I had to give this option a name, it'd definitely have "fragsize" in its name to link it's meaning to the other parameter.  Perhaps `hl.fragsizeIsMinimum` inverted meaning.  Sadly `hl.fragsize` is not camel cased; ah well.  

Maybe these should be _secret_ options with code comments about filing a JIRA issue if you want to change them :-).  After all... I'm not sure _all_ options need to be documented.

Reply:
> hl.closestTo ... I wish it simply always worked this way...

I'm guessing whoever wrote the original UH decided against it for performance reasons. There is a  [comment](https://github.com/apache/lucene-solr/blob/e06ad4cfb587e1bb69a80c0b531df0e52c2da89b/lucene/highlighter/src/java/org/apache/lucene/search/uhighlight/LengthGoalBreakIterator.java#L149) to suggest that. Also note that depending on which index is closer the BI may had to invoke `following()` one extra time inside `moveToBreak()`. I think `moveToBreak()` is only necessary when a default-summary is being made and I can remove it from the code when a match is being highlighted.

The performance hit would be depending on the fragment's length that contains the goal index. In WORD mode I'd hope this is negligible. It will necessarily be worse for SENTENCE.

You've brought up a new point to keep in mind: simplifying the options. I'll be back with updated code that we can discuss further.